### PR TITLE
metric: first meaningful paint 1.0

### PR DESCRIPTION
**Pull request by paulirish**
At long last, we have the first meaningful paint metric.

This follows the algorithm defined in [Time to First Meaningful Paint: a layout-based approach](https://docs.google.com/document/d/1BR94tJdZLsin5poeet0XoTW60M0SjvOJQttKT-JK8HI/edit#).

Using `--save-trace` I've spent some time verifying these values. They are looking mighty fine. :)

### CODE REVIEW

**Comment by paulirish** [comment]
Build is failing because of the metrics tests. 

I'm going to rebase this against #319 when its ready and fix the tests in that hit.

**Comment by paulirish** [comment]
This is now rebaselined against #319, which should land first.

**Comment by paulirish** [comment]
Test failures are only FMP stuff. 
If the direction of this approach looks good to folks, I'll sort out the tests next.

File: src/audits/performance/first-meaningful-paint.js [hunk 1]
```diff
@@ -51,48 +58,187 @@ class FirstMeaningfulPaint extends Audit {
...
-          });
-        });
+    if (!artifacts.traceContents || !Array.isArray(artifacts.traceContents)) {
+      throw new Error(FAILURE_MESSAGE);
+    }
+
+    try {
+      const evts = this.collectEvents(artifacts.traceContents);
+
+      /* eslint-disable no-multi-spaces  */
```
Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: I don't think you use that here

File: src/audits/performance/first-meaningful-paint.js [hunk 2]
```diff
@@ -51,48 +58,187 @@ class FirstMeaningfulPaint extends Audit {
...
+    }
+  }
+
+  static calculateScore(data) {
+    // there are a few candidates for fMP:
+    // * firstContentfulPaint: the first time that text or image content was painted.
+    // * fMP basic: paint after most significant layout
+    // * fMP page height: basic + scaling sigificance to page height
+    // * fMP webfont: basic + waiting for in-flight webfonts to paint
+    // * fMP full: considerig both page height + webfont heuristics
```
Comment (brendankenny) [inline src/audits/performance/first-meaningful-paint.js]: considering

**Comment by brendankenny** [comment]
tests fixed. Should be good to go.

**Pull request merged**